### ISSUE TITLE
Application in Physics Simulation

### ISSUE THREAD

**Issue by crsiper** [zju3dv/GVHMR#24]
Hi, thank you for this incredible work!
How can I generate the required .pkl file for PHC from the output .pt file? 
I'm trying to apply the resulting motion sequence to physics simulation in IsaacGym using PHC. I appreciate your help and look forward to your response!

**Comment by kexul** [comment]
See here: https://github.com/zju3dv/GVHMR/issues/17

**Comment by crsiper** [comment]
> See here: #17
Hi, I have read the comment.
```
verts = verts[..., [0, 2, 1]]
verts[..., 1] *= -1
min_z = verts[:, :, 2].min(1).values.numpy()  # (L)
degree = 1
coefficients = np.polyfit(np.arange(len(min_z)), min_z, degree)
polynomial = np.poly1d(coefficients)
fitted_values = polynomial(np.arange(len(min_z)))
verts[:, :, 2] -= torch.from_numpy(fitted_values)[:, None].to(verts)
```
What does the variable "verts" refer to? Is it the "smpl_params_global":"body_pose" loaded from the .pt file?

**Comment by kexul** [comment]
> > See here: #17
> > Hi, I have read the comment.
> 
> ```
> verts = verts[..., [0, 2, 1]]
> verts[..., 1] *= -1
> min_z = verts[:, :, 2].min(1).values.numpy()  # (L)
> degree = 1
> coefficients = np.polyfit(np.arange(len(min_z)), min_z, degree)
> polynomial = np.poly1d(coefficients)
> fitted_values = polynomial(np.arange(len(min_z)))
> verts[:, :, 2] -= torch.from_numpy(fitted_values)[:, None].to(verts)
> ```
> 
> What does the variable "verts" refer to? Is it the "smpl_params_global":"body_pose" loaded from the .pt file?

I don't know. Actually I've tested the code and it does not work. 

In my practice, I manually aligned the pose with the Amass pose (which is used in PHC in their `convert_data_smpl.py` script ) in blender, record the pelvis rotation and replace the `global_orient` with it.
Then I inverted the x and z axis of the `trans`. 
Now the simulation CAN run, but the result I get so far is not very good. (I'm using simulation to fix some foot sliding and penetration issue, but It's not working...

**Comment by crsiper** [comment]
Yes. I also try to use `convert_data_smpl.py` script in PHC to convert the data, but the shape of `"body_pose"` is (total frames, 63) which means it has 21 joints, not match with the standard smpl parameters with 24 joints. Also in `convert_data_smpl.py` the variable `"pose_aa"` needs to be shape of (total frames, 24, 3) so that `"body_pose"` cant be used directly here. I am currently try to figure out how to transfer these data.

**Comment by kexul** [comment]
> Yes. I also try to use `convert_data_smpl.py` script in PHC to convert the data, but the shape of `"body_pose"` is (total frames, 63) which means it has 21 joints, not match with the standard smpl parameters with 24 joints. Also in `convert_data_smpl.py` the variable `"pose_aa"` needs to be shape of (total frames, 24, 3) so that `"body_pose"` cant be used directly here. I am currently try to figure out how to transfer these data.

If I understand correctly, the body pose is 21 joints, the global orient is pelvis joint. You can just concat them and use it in PHC. Two hand joints are not presented in the gvhmr result and not used in PHC, if you check the `convert_data_smpl.py` code, only the first 22 joints are used:
https://github.com/ZhengyiLuo/PHC/blob/d0c1cf090b3979584871799b3b3d65e905430af4/scripts/data_process/convert_data_smpl.py#L86

You can check my code, which transfers the data: 
```python
import pickle
import numpy as np 

with open('hmr4d_results.pkl', 'rb') as f:
    result = pickle.load(f)
    result = result['smpl_params_global']

pose = np.concatenate([result['global_orient'], result['body_pose']], axis=1)
# If you get the corrected rotation manually in blender, you can use that 
# corrected_orient = np.array([1.31, 1.25, 0.83] * result['global_orient'].shape[0])
# pose = np.concatenate([corrected_orient, result['body_pose']], axis=1)
trans = result['transl']
# I inverted the x axis and z axis movement because I noticed character are move backward after import.  
# trans[..., 0] *= -1
# trans[..., 2] *= -1


phc_data = {}
phc_data['pose_aa'] = pose
phc_data['betas'] = result['betas']
phc_data['trans'] = trans
joblib.dump({'phc_data':phc_data}, 'phc_data.pkl')
```

**Comment by crsiper** [comment]
> > Yes. I also try to use `convert_data_smpl.py` script in PHC to convert the data, but the shape of `"body_pose"` is (total frames, 63) which means it has 21 joints, not match with the standard smpl parameters with 24 joints. Also in `convert_data_smpl.py` the variable `"pose_aa"` needs to be shape of (total frames, 24, 3) so that `"body_pose"` cant be used directly here. I am currently try to figure out how to transfer these data.
> 
> If I understand correctly, the body pose is 21 joints, the global orient is pelvis joint. You can just concat them and use it in PHC. Two hand joints are not presented in the gvhmr result and not used in PHC, if you check the `convert_data_smpl.py` code, only the first 22 joints are used: https://github.com/ZhengyiLuo/PHC/blob/d0c1cf090b3979584871799b3b3d65e905430af4/scripts/data_process/convert_data_smpl.py#L86
> 
> You can check my code, which transfers the data:
> 
> ```python
> import pickle
> import numpy as np 
> 
> # load the result in the pt file
> result = result['smpl_params_global']
> 
> pose = np.concatenate([result['global_orient'], result['body_pose']], axis=1)
> # If you get the corrected rotation manually in blender, you can use that 
> # corrected_orient = np.array([1.31, 1.25, 0.83] * result['global_orient'].shape[0]).reshape(-1, 3)
> # pose = np.concatenate([corrected_orient, result['body_pose']], axis=1)
> trans = result['transl']
> # I inverted the x axis and z axis movement because I noticed character are moving backward after import.  
> # trans[..., 0] *= -1
> # trans[..., 2] *= -1
> 
> 
> phc_data = {}
> phc_data['pose_aa'] = pose
> phc_data['betas'] = result['betas']
> phc_data['trans'] = trans
> joblib.dump({'phc_data':phc_data}, 'phc_data.pkl')
> ```

Thank you so much for the explanation! Very helpful. I have successfully transferred the data by the running following code:
```
import torch
import pickle
import numpy as np

hmr4d_results_path = 'gvhmr_data/before_converted/model_generate/hmr4d_results.pt'

result = torch.load(hmr4d_results_path, weights_only=True)
result = result['smpl_params_global']

pose = np.concatenate([result['global_orient'], result['body_pose']], axis=1) 
trans = result['transl'].cpu().numpy()
betas = result['betas'].cpu().numpy()

phc_data = {}
phc_data['smpl_params'] = {}
phc_data['smpl_params']['pose_aa'] = pose
phc_data['smpl_params']['betas'] = betas
phc_data['smpl_params']['trans'] = trans

output_path = 'gvhmr_data/before_converted/phc_data.pkl'
with open(output_path, 'wb') as f:
    pickle.dump(phc_data, f, protocol=pickle.HIGHEST_PROTOCOL)
```

Then I run the script `convert_data_smpl.py` in PHC to get to final .pkl file. But when I run the phc_data.pkl with PHC in Isaac Gym, the joint remapping was fine except there was something wrong with the gravity axis. The keypoints were aligned horizontally instead of vertically, as shown in the picture:
![Screenshot from 2024-09-25 15-17-01](https://github.com/user-attachments/assets/f11f4181-6c35-4d86-aca7-b7347f1e09bb)

I tried to rotate the axis to match the simulation's as mentioned in [https://github.com/zju3dv/GVHMR/issues/17](url), replacing the variable `verts` with `pose`, and in order to change the order of xyz axis in `pose`, I reshaped  `pose` like this: 
```
frames = pose.shape[0]
pose = pose.reshape(frames, 22, 3)
pose = pose[:, :, [1,0,2]] 
pose[:,:, 0] *= -1 
pose = pose.reshape(frames, -1)
```
Then I regenerate the phc_data.pkl. However, when I ran it again in the simulation, it seemed like the joint remapping was incorrect as the movement was very unstable. 

Is it the problem of reshape? How can I fix the gravity axis correctly?

**Comment by kexul** [comment]
That's what I've encountered these days! 😃 As I said earlier, I rotated the pelvis in blender manually to mach the axis of AMASS data. ( You can check this https://github.com/nghorbani/amass/blob/master/support_data/github_data/amass_sample.npz, load it in blend by installing the SMPL-X addon and then `Add Animation`.)

I found that [1.382, 1.279, 1.106] fit most of my pose generated by GVHMR, you can try these magic numbers too. 😆  
Here is my code:
```python
orient = [1.382, 1.279, 1.106]
orient = np.array(orient * new_res['global_orient'].shape[0]).reshape(-1, 3)
pose = np.concatenate([ori, new_res['body_pose']], axis=1)
phc_data['pose_aa'] = pose
phc_data['trans'] = result['transl'][..., [2, 0, 1]]
```

**Comment by crsiper** [comment]
> That's what I've encountered these days! 😃 As I said earlier, I rotated the pelvis in blender manually to mach the axis of AMASS data. ( You can check this https://github.com/nghorbani/amass/blob/master/support_data/github_data/amass_sample.npz, load it in blend by installing the SMPL-X addon and then `Add Animation`.)
> 
> I found that [1.382, 1.279, 1.106] fit most of my pose generated by GVHMR, you can try these magic numbers too. 😆 Here is my code:
> 
> ```python
> orient = [1.382, 1.279, 1.106]
> orient = np.array(orient * new_res['global_orient'].shape[0]).reshape(-1, 3)
> pose = np.concatenate([ori, new_res['body_pose']], axis=1)
> phc_data['pose_aa'] = pose
> phc_data['trans'] = result['transl'][..., [2, 0, 1]]
> ```

These numbers did work in my simulation as well, it has been a great help and appreciate a lot! Now the keypoints are normal, but the smpl robot can not quite catch those keypoints. I will try the method offered by author in issue #17 to increase stability later.
Blender is not in part of my project plan so I did not apply the result in blender. I wish I could discuss more with you about that. I might try that later!

**Comment by kexul** [comment]
> Now the keypoints are normal, but the smpl robot can not quite catch those keypoints. 

Yeah, far worse than the result posted by the repo owner. 
My result:
https://github.com/user-attachments/assets/5646fe5e-f366-4686-8e03-f679d8267148

Maybe the `fitting a pseudo-ground plane to the z-values can further improve stability` trick would help, but I don't know how to implement.
